kore/src/Kore/Repl.hs

```diff
@@ -81,39 +84,39 @@ runRepl
     -- ^ optional output file
     -> m ()
 runRepl axioms' claims' logger replScript replMode outputFile = do
-    mNewState <- evaluateScript replScript
+    (newState, _) <-
+            (\rwst -> execRWST rwst config state)
+            $ evaluateScript replScript
     case replMode of
         Interactive -> do
             replGreeting
-            evalStateT (forever repl0) (maybe state id mNewState)
-        RunScript ->
-            case mNewState of
-                Nothing -> liftIO exitFailure
-                Just newState -> do
-                    runReplCommand ProofStatus newState
-                    runReplCommand Exit newState
+            flip evalStateT newState
+                $ flip runReaderT config
+                $ forever repl0
+        RunScript -> do
+            runReplCommand ProofStatus newState
+            runReplCommand Exit newState
 
   where
 
-    runReplCommand :: ReplCommand -> ReplState claim m -> m ()
+    runReplCommand :: ReplCommand -> ReplState claim -> m ()
     runReplCommand cmd st =
-        void $ evalStateT (replInterpreter printIfNotEmpty cmd) st
+        void
+            $ flip evalStateT st
+            $ flip runReaderT config
+            $ replInterpreter printIfNotEmpty cmd
 
-    evaluateScript :: ReplScript -> m (Maybe (ReplState claim m))
-    evaluateScript rs =
-        maybe
-            (pure . pure $ state)
-            (parseEvalScript state)
-            (unReplScript rs)
+    evaluateScript :: ReplScript -> RWST (Config claim m) String (ReplState claim) m ()
+    evaluateScript = maybe (pure ()) parseEvalScript . unReplScript
 
-    repl0 :: StateT (ReplState claim m) m ()
+    repl0 :: ReaderT (Config claim m) (StateT (ReplState claim) m) ()
```

Comment:
Why do we have a `ReaderT _ (StateT _ m)` stack in some places, but use `RWST` in others? I recall there was a reason we used `StateT` instead of `RWST` in some places, but I don't remember what it was. Anyway, to avoid some of the awkwardness of chaining `runReaderT` with `runStateT` you could also use `RWST (Config claim m) () (ReplState claim) m` because `()` is a `Monoid`.

Reply:
Would `RWST _ () _` have the `WriterT` performance problems if we use `()` as the write type? I wasn't sure, which is why I used the `ReaderT ... StateT`. See a few lines up, the `forever repl0`, which is the worrying bit.